File: sol/solution/0001-0100/0006. Zigzag Conversion/Solution.ex
defmodule Solution do
  @spec convert(s :: String.t, num_rows :: integer) :: String.t
  def convert(s, 1), do: s
  def convert(s, num_rows) do
    cycle_len = 2 * num_rows - 2
    0..(num_rows - 1)
    |> Enum.map(fn row -> add_row(s, num_rows, cycle_len, row) end)
    |> IO.iodata_to_binary()
  end
  def add_row(s, num_rows, cycle_len, row), do: add_row(s, num_rows, cycle_len, row, 0, [])
  def add_row(s, num_rows, cycle_len, row, cycle, acc)
      when row + cycle * cycle_len < byte_size(s) do
    index = row + cycle * cycle_len
    acc = [:binary.at(s, index) | acc]
    acc =
      if row > 0 and row < num_rows - 1 and (cycle + 1) * cycle_len - row < byte_size(s) do
        index = (cycle + 1) * cycle_len - row
        [:binary.at(s, index) | acc]
      else
        acc
      end
    add_row(s, num_rows, cycle_len, row, cycle + 1, acc)
  end
  def add_row(_, _, _, _, _, acc), do: Enum.reverse(acc)
end

File: sol/solution/0001-0100/0006. Zigzag Conversion/Solution.erl
-spec convert(S :: unicode:unicode_binary(), NumRows :: integer()) -> unicode:unicode_binary().
convert(S, NumRows) when NumRows =< 1; NumRows >= byte_size(S) ->
    S;
convert(S, NumRows) ->
    Rows = lists:duplicate(NumRows, []),
    convert(S, NumRows, 1, 1, Rows).
convert(<<>>, _NumRows, _CurRow, _Dir, Rows) ->
    unicode:characters_to_binary(lists:flatten(lists:map(fun lists:reverse/1, Rows)));
convert(<<C/utf8, Rest/binary>>, NumRows, CurRow, Dir, Rows) ->
    NewRows = add_char(Rows, CurRow, C),
    NewDir = case {CurRow, Dir} of
        {1, -1} -> 1;
        {NumRows, 1} -> -1;
        _ -> Dir
    end,
    NewCurRow = CurRow + NewDir,
    convert(Rest, NumRows, NewCurRow, NewDir, NewRows).
add_char(Rows, CurRow, C) ->
    {NewRows, _} = lists:mapfoldl(
        fun(Row, Idx) ->
            if Idx =:= CurRow -> {[C | Row], Idx + 1};
               true -> {Row, Idx + 1}
            end
        end, 1, Rows),
    NewRows.